### DISCUSSION TITLE
How to set up MK22FN512VLH12 clock for crystalless operation?

### DISCUSSION THREAD

**Discussion by mondalaci** [zephyrproject-rtos/zephyr#64815] [Q&A] [answered]
I can build the blinky example Zephyr application, flash it to the FRDM-K22F dev board, and it works fine.

However, when flashing the application on our custom board (with a modified LED GPIO according to our schematic), it doesn't blink its LED, and debugging it revealed that it stops upon configuring the clock. It's not surprising since our custom board doesn't have a crystal, unlike the FRDM-K22F board.

Before starting to use Zephyr, we've been using a bare metal firmware, and you can see [its clock configuration code](https://github.com/UltimateHackingKeyboard/firmware/blob/9f5c27b7adc08aa0a93c56752630087327b2dfc2/right/src/init_clock.c). I tried to replicate this clock configuration with Zephyr but couldn't so far.

In Kconfig, I set:

`CONFIG_SOC_SERIES_KINETIS_K2X=y`

As for the devicetree, I found [the relevant bindings](https://docs.zephyrproject.org/latest/build/dts/api/bindings/clock/nxp,kinetis-sim.html), and set the following based on our bare metal firmware:

```
&sim {
    pllfll-select = <KINETIS_SIM_PLLFLLSEL_MCGPLLCLK>;
    er32k-select = <KINETIS_SIM_ER32KSEL_OSC32KCLK>;
    clkout-source = <KINETIS_SIM_CORESYS_CLK>;
};
```

The firmware builds without the `clkout-source` property, but with it, I get the following error:

```
/home/laci/projects/firmware/deps/zephyr/drivers/clock_control/clock_control_mcux_sim.c:87:12: error: 'SIM_Type' has no member named 'CHIPCTL'
   87 |         SIM->CHIPCTL = (SIM->CHIPCTL & ~SIM_CHIPCTL_CLKOUTSEL_MASK)
      |            ^~
/home/laci/projects/firmware/deps/zephyr/drivers/clock_control/clock_control_mcux_sim.c:87:28: error: 'SIM_Type' has no member named 'CHIPCTL'
   87 |         SIM->CHIPCTL = (SIM->CHIPCTL & ~SIM_CHIPCTL_CLKOUTSEL_MASK)
      |                            ^~
/home/laci/projects/firmware/deps/zephyr/drivers/clock_control/clock_control_mcux_sim.c:87:41: error: 'SIM_CHIPCTL_CLKOUTSEL_MASK' undeclared (first use in this function); did you mean 'SIM_SOPT2_CLKOUTSEL_MASK'?
   87 |         SIM->CHIPCTL = (SIM->CHIPCTL & ~SIM_CHIPCTL_CLKOUTSEL_MASK)
      |                                         ^~~~~~~~~~~~~~~~~~~~~~~~~~
      |                                         SIM_SOPT2_CLKOUTSEL_MASK
/home/laci/projects/firmware/deps/zephyr/drivers/clock_control/clock_control_mcux_sim.c:87:41: note: each undeclared identifier is reported only once for each function it appears in
/home/laci/projects/firmware/deps/zephyr/drivers/clock_control/clock_control_mcux_sim.c:88:19: warning: implicit declaration of function 'SIM_CHIPCTL_CLKOUTSEL'; did you mean 'SIM_SOPT2_CLKOUTSEL'? [-Wimplicit-function-declaration]
   88 |                 | SIM_CHIPCTL_CLKOUTSEL(NXP_KINETIS_SIM_CLKOUT_SOURCE);
      |                   ^~~~~~~~~~~~~~~~~~~~~
      |                   SIM_SOPT2_CLKOUTSEL
ninja: build stopped: subcommand failed.
```

Then I checked whether `SIM->CHIPCTL` should exist. For example, [it's present in MKE12Z7.h](https://github.com/zephyrproject-rtos/hal_nxp/blob/3731aefd0cd55f8507498c336b3256268ecd9d61/mcux/mcux-sdk/devices/MKE12Z7/MKE12Z7.h#L12956), but it's not present in [MK22F51212.h](https://github.com/zephyrproject-rtos/hal_nxp/blob/3731aefd0cd55f8507498c336b3256268ecd9d61/mcux/mcux-sdk/devices/MK22F51212/MK22F51212.h), and the latter is relevant for the MK22FN512VLH12 MCU used.

The above begs the question: Am I missing something, or is this a bug?

**Comment by DerekSnell** [comment, marked as answer]
Hi @mondalaci ,
The [clock_init()](https://github.com/zephyrproject-rtos/zephyr/blob/ec527228613e2768caaddedaa5cdea04951d3a06/soc/arm/nxp_kinetis/k2x/soc.c#L85C27-L85C43) used by the K22F device does not have an option to clock from the internal oscillator.  And it appears it is not configured by devicetree.  If you want to change the clock config for your board, you will need to modify this function.

And FYI, for the devices that do have the `SIM_CHIPCTL_CLKOUTSEL` bitfield, those bits control a CLKOUT pin.  This is a mux that selects what clock will be driven externally on that pin.  But it does not configure the internal SOC clocks.  Since you said "The firmware builds without the clkout-source property", I would leave that property out of your devicetree.

Best regards

**Comment by mondalaci** [comment]
@DerekSnell Thank you very much for your help and clarification!

I've been trying to figure out how to modify `clock_init()` to replicate my crystalless KSDK clock configuration in Zephyr, so far, unsuccessfully. I can see that the board configuration and Kconfig also have relevant settings, and my understanding of Kinetis clocks is also incomplete.

I could use some expert help, and I'd be happy to pay for it. I probably only need some hand-holding, and I can figure out the rest.

Your further help and recommendations of experts would be much appreciated.

**Comment by DerekSnell** [comment]
Hi @mondalaci ,
This GitHub site is focused on Zephyr, and the NXP members monitoring this site are focused on enabling and supporting Zephyr on our SOCs.  But if you need help configuring SOC hardware, like configuring the SOC to use a specific clock source, and the question is not specific to Zephyr, then I recommend using [NXP's Community forum](https://community.nxp.com/).  There are forums for different products, and this is the [Kinetis Community](https://community.nxp.com/t5/Kinetis-Microcontrollers/bd-p/kinetis).  You can search that Community for existing answers.  Or post a new question there, and it will be assigned to a support engineer focused on that SOC or feature.

For paid support, these are some options that come to mind:

- [NXP Professional Support](https://www.nxp.com/support/support/nxp-engineering-services/professional-support-for-processors-and-microcontrollers:PREMIUM-SUPPORT) - support hours can be purchased
- [NXP Partners](https://www.nxp.com/design/partner-marketplace:PARTNER-MARKETPLACE) - this site lists 3rd party partners who work closely with NXP.  You can search for consultants with experience related to your needs.
- [Zephyr Project ecosystem vendors](https://www.zephyrproject.org/ecosystem-vendor-offerings/) - lists partners with Zephyr experience.  You can filter for consultants.

I hope that helps, best regards
